### DISCUSSION TITLE
Validation of string byte-length?

### DISCUSSION THREAD

**Discussion by joaohaas** [json-schema-org/community#377] [Q&A] [answered]
I want to ensure a specific string field that carries textual data does not exceed a specific byte-length. Since strings are UTF-8 encoded, `maxLength` will not really validate this correctly. I tried changing `contentEncoding` to `binary` as well, but that didn't seem 
to change the behavior of `maxLength`.

For reference, I would like this schema:

```json
{
  "$schema": "https://json-schema.org/draft/2019-09/schema",
  "type": "string",
  "maxLength": 4,
}
```

to fail against this JSON (since the string has 8 bytes):

```json
"ãããã"
```

Is there any way to validate this? If not, is it on the roadmap for future json schema releases?

**Comment by jdesrosiers** [comment, marked as answer]
Unfortunately, there's no way to validate this with JSON Schema. I've always thought that the way this keyword is defined (counting visible characters) made it practically useless. Almost always, the reason to constrain the length of a string in a schema is to ensure that it will fit in some space in memory (usually a cell in a relational database table). I'd even call `maxLength` a foot-gun because people use this keyword all the time to check that their strings will fit in their database tables and don't realize that it won't always work.

I'd support adding a new keyword that constrains the maximum size of strings.

**Reply by joaohaas** [reply]
Yeah, I thought so, just wanted to make sure. I'll just use a custom validator for now, but yeah, it sure would be nice to have that as a 'standard' keyword.

What is the appropriate channel to suggest this feature for a future spec?

**Reply by jdesrosiers** [reply]
> What is the appropriate channel to suggest this feature for a future spec?

You can start a new discussion with a specific proposal for the new keyword(s). If it's agreed that it should be added, someone will write it up and add it to the spec.

**Reply by karenetheridge** [reply]
I could definitely see a vocabulary dedicated to unicode properties of strings, or things related to utf8 encoding.  I've occasionally had the desire to express that an array is sorted, for some value of sorted (e.g. a particular unicode collation for strings), or that a certain string value is less than a value in the schema (again using a particular unicode collation to do the string comparison).

**Reply by SorinGFS** [reply]
@jdesrosiers 

> I'd support adding a new keyword that constrains the maximum size of strings

If such a word were to be introduced, it should be called `maxBytes`, and it should not only concern strings but also numbers. In fact this keyword would be more useful for numbers than strings since strings can benefit in most fo the cases of json string being utf-8, while for float numbers there is no option.

**Comment by karenetheridge** [comment]
`contentEncoding: binary` should work with `maxLength: ..`, provided you actually do encode your string as binary (that is, it will end up encoded twice in json, but then when it is inflated from json to a data structure native to your application, you'll get a binary string back). But if it's just a string in the json data, you won't get the right outcome.
